### PR TITLE
1.5 enhancer

### PR DESCRIPTION
**Pull request by arschmitz**
A new stand alone auto enhancment module for declarative widgets.

### CODE REVIEW

File: js/widgets/enhancer.backcompat.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+			return prototype.initSelector ||
+				$[ prototype.namespace ][ prototype.widgetName ].prototype.initSelector ||
+				originalGenerator( prototype, ns );
+		};
+
+	$.fn.enhance._filter = filter;
+	$.fn.enhance.initGenerator = generator;
+
+}
+});
```
Comment (arschmitz) [inline js/widgets/enhancer.backcompat.js]: missing new line

File: js/widgets/enhancer.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+				options[ option ] = value;
+			}
+		}
+
+		return options;
+	}
+});
+
+return plugin;
+} );
```
Comment (arschmitz) [inline js/widgets/enhancer.js]: missing new line

File: js/widgets/enhancer.backcompat.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	}
+})( function( $ ) {
+if ( $.mobileBackcompat !== false ) {
+	var originalGenerator = $.fn.enhance.initGenerator,
+		filter = function( elements ) {
+			elements = elements.not( $.mobile.keepNative );
+
+			if ( $.mobile.ignoreContentEnabled ) {
+				elements.each( function() {
+					if ( $( this ).closest( "[data-" + $.mobile.ns + "enhanced='false']" ).length ) {
```
Comment (gabrielschulhof) [inline js/widgets/enhancer.backcompat.js]: I think the name of the attribute is supposed to be ```data-enhance```, not ```data-enhanced```

Comment (gabrielschulhof) [inline js/widgets/enhancer.backcompat.js]: https://github.com/jquery/jquery-mobile/blob/master/js/helpers.js#L142

File: js/widgets/widget.backcompat.js [hunk 4]
```diff
@@ -42,6 +42,7 @@ if ( $.mobileBackcompat !== false ) {
 		_create: function() {
 			this._setInitialOptions();
 			this._super();
+
```
Comment (gabrielschulhof) [inline js/widgets/widget.backcompat.js]: Is it worth touching this file for only one blank line?

Comment (arschmitz) [inline js/widgets/widget.backcompat.js]: there were other changes this is the results of a rebase that already incorporated them ( apparently minus this one line change )

File: tests/integration/forms/index.html [hunk 5]
```diff
@@ -84,7 +84,7 @@
 				</div>
 				<div data-nstest-role="fieldcontain">
 					<label for="testRange">Test Range</label>
-					<input id="testRange" type="range" min="0" max="100" value="42"></input>
+					<input id="testRange" type="number" data-nstest-role="slider" min="0" max="100" value="42"></input>
```
Comment (arschmitz) [inline tests/integration/forms/index.html]: revert this

File: js/widgets/widget.backcompat.js [hunk 6]
```diff
@@ -44,6 +44,7 @@ if ( $.mobileBackcompat !== false ) {
 				this._super();
 				if ( !this.options.enhanced && this.options.wrapperClass ) {
 					this._addClass( this.widget(), null, this.options.wrapperClass );
+
```
Comment (arschmitz) [inline js/widgets/widget.backcompat.js]: revert this

File: tests/unit/slider/index.html [hunk 7]
```diff
@@ -37,32 +37,32 @@
 
 	<div id="foo" data-nstest-role="page">
 		<div data-nstest-role="fieldcontain">
-			<input type="range" name="slider-1" id="range-slider-up" value="0" min="0" max="100" data-nstest-theme="b" data-nstest-track-theme="a" />
+			<input type="number" data-nstest-role="slider" name="slider-1" id="range-slider-up" value="0" min="0" max="100" data-nstest-theme="b" data-nstest-track-theme="a" />
```
Comment (arschmitz) [inline tests/unit/slider/index.html]: revert all these

**Comment by arschmitz** [comment]
Still some usages of `.trigger( "create" )` in demos that need to be cleaned up

**Comment by arschmitz** [comment]
We had talked about removing our usage of custom `initSelectors` as part of this. This PR does not do that though. Maybe i should add that and move all the custom ones to backCompat before we land this. That does involve a lot of work for the demos and tests though...

File: js/widgets/enhancer.backcompat.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
+/*!
+ * jQuery Mobile Enhnacer @VERSION
```
Comment (gabrielschulhof) [inline js/widgets/enhancer.backcompat.js]: Enhancer

File: js/widgets/enhancer.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
+/*!
+ * jQuery Mobile Enhnacer @VERSION
```
Comment (gabrielschulhof) [inline js/widgets/enhancer.js]: Enhancer

File: js/widgets/enhancer.backcompat.js (hunk 8 above)
Comment (gabrielschulhof) [inline js/widgets/enhancer.backcompat.js]: "Enhancer Backcompat" maybe so it's not the same as the actual enhancer.

File: tests/unit/degrade-inputs/degradeInputs.js [hunk 10]
```diff
@@ -3,9 +3,79 @@
  */
 
 ( function( $ ) {
-module( 'jquery.mobile.degradeInputs.js' );
+module( "jquery.mobile.degradeInputs.js" );
 
-test( 'keepNative elements should not be degraded', function() {
-	deepEqual( $( 'input#not-to-be-degraded' ).attr( "type" ), "range" );
+test( "degrdeInputs works on page init", function() {
```
Comment (gabrielschulhof) [inline tests/unit/degrade-inputs/degradeInputs.js]: degrade

File: tests/unit/degrade-inputs/degradeInputs.js [hunk 11]
```diff
@@ -3,9 +3,79 @@
...
+
+	strictEqual( $( "#degrade-range" ).attr( "type" ), "number",
+		"Range inputs are degrade to type number" );
+	strictEqual( $( "#degrade-search" ).attr( "type" ), "text",
+		"Search inputs degrade to type text" );
+	strictEqual( $( "#degrade-color" ).attr( "type" ), "color",
+		"setting an input type to false cause no degradation of that type" );
+} );
+
+test( "degrdeInputs works on page enhance", function() {
```
Comment (gabrielschulhof) [inline tests/unit/degrade-inputs/degradeInputs.js]: degrade

**Comment by arschmitz** [comment]
landed in 1.5-dev

**Pull request closed without merging**